Zend/zend_hash.c

```diff
@@ -2362,15 +2654,29 @@ ZEND_API zend_result ZEND_FASTCALL zend_hash_move_forward_ex(HashTable *ht, Hash
 
 	idx = _zend_hash_get_valid_pos(ht, *pos);
 	if (idx < ht->nNumUsed) {
-		while (1) {
-			idx++;
-			if (idx >= ht->nNumUsed) {
-				*pos = ht->nNumUsed;
-				return SUCCESS;
+		if (HT_IS_PACKED(ht)) {
+			while (1) {
+				idx++;
+				if (idx >= ht->nNumUsed) {
```

Comment:
I assume zend_hash_move_forward_ex is used pretty frequently.

can `*pos = idx;` and `*pos = ht->nNumUsed` be combined? I assume the resulting assembly code would be smaller

The above statements `idx < ht->nNumUsed` and `idx++` make it look like idx and ht->nNumUsed will be the identical in this branch.

i.e. 

```php
				if (idx >= ht->nNumUsed || Z_TYPE(ht->arPacked[idx]) != IS_UNDEF) {
					*pos = idx;
					ZEND_ASSERT(idx <= ht->nNumUsed);
					return SUCCESS;
```

Same for the non-packed case

Reply:
In general, ``ht->nNumUsed`` may be changed between iterations (some elements may be added-removed). I can't be sure, if this optimization is safe for all cases. I prefer to keep the existing logic.